### PR TITLE
Nukes all embedded images for a Workitem.

### PR DESCRIPTION
**Pull request by lremedi**
Param: workitem number

### CODE REVIEW

**Comment by lremedi** [comment]
Important: I was not able to use built-in sql functionality for xml because html is not well-formed and throwing errors  when trying to parse, for example IMG never has a closing tag or being self-closed

File: one-off/nuke-workitem-embedded-images.sql [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    if (@@rowcount <> 1)
+    begin
+        raiserror('Invalid Workitem number', 0, 1)
+        return
+    end
+
+    declare @blankimage varbinary(max) = 0x89504e470d0a1a0a0000000d4948445200000001000000010100000000376ef9240000001049444154789c626001000000ffff03000006000557bfabd40000000049454e44ae426082
+
+    update Blob 
+    set Value = @blankimage
```
[the remark this answers is not in the document]
Comment (pmamut) [reply one-off/nuke-workitem-embedded-images.sql]: `Blob.ContentType` is meant for ["hashed" BLOB URLs](https://github.com/versionone/Core/blob/23.3/VersionOne.Web/Tools/AttachmentHandler.cs#L315)

`EmbeddedImage_Now.ContentType` is the droid we're looking for when dealing with proper embedded images

File: one-off/nuke-workitem-embedded-images.sql [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        ) as imgelem,
+        convert(binary(20), '0x'+ substring(
+            @richtextcol,
+            charindex('downloadblob.img/', @richtextcol, startpos) + len('downloadblob.img/'),
+            charindex('"', @richtextcol, charindex('downloadblob.img/', @richtextcol, startpos) + len('downloadblob.img/')) - charindex('downloadblob.img/', @richtextcol, startpos) - len('downloadblob.img/')
+        ),1) as hash
+    from
+        recursivecte
+    where
+        charindex('downloadblob.img/', @richtextcol, startpos) > 0
```
Comment (pmamut) [inline one-off/nuke-workitem-embedded-images.sql:54]: i think you can just look for `downloadblob.img/` as the starting marker, and the immediately following `"` as the trailing marker for the hash, without bothering with it being in an ``

File: one-off/nuke-workitem-embedded-images.sql [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    begin
+        set @RichTextCol = stuff(
+            @RichTextCol,
+            @StartIndex,
+            charindex('">', substring(@RichTextCol, @StartIndex, len(@RichTextCol))) + 2,
+            @BlankBase64Image
+        );
+
+        set @StartIndex = charindex('<img src="data:image/', @RichTextCol, @StartIndex + len(@BlankBase64Image));
+    end;
```
Comment (pmamut) [inline one-off/nuke-workitem-embedded-images.sql:76]: ... same with `src="data:image/` and its immediately following `"` for inline images

also, doesn't this currently nuke the entire `` with `@BlankBase64Image` value? does the latter contain the entire replacement `` tag?

File: one-off/nuke-workitem-embedded-images.sql [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	join dbo.OidSplit(@oid) oid on ba.ID = oid.ID and ba.AssetType = oid.AssetType
+
+    if (@@rowcount <> 1)
+    begin
+        raiserror('Invalid asset oid', 0, 1)
+        return
+    end
+
+    declare @blankimage varbinary(max) = 0x89504e470d0a1a0a0000000d4948445200000001000000010100000000376ef9240000001049444154789c626001000000ffff03000006000557bfabd40000000049454e44ae426082
+	declare @blankImageBase64 varchar(max) = CAST(N'' AS XML).value('xs:base64Binary(xs:hexBinary(sql:variable("@blankimage")))', 'VARCHAR(MAX)')
```
Comment (pmamut) [inline one-off/nuke-workitem-embedded-images.sql:103]: so this?
```suggestion
	declare @blankImageBase64 varchar(max) = '<img src="data:image/' + CAST(N'' AS XML).value('xs:base64Binary(xs:hexBinary(sql:variable("@blankimage")))', 'VARCHAR(MAX)') + '">'
```

File: one-off/nuke-workitem-embedded-images.sql [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    exec dbo._SaveString N'img/png', @imgpng output
+
+    update Blob 
+    set Value = @blankimage, ContentType=@imgpng
+    where ID in (
+        select Content
+        from dbo.EmbeddedImage
+        where AssetID = @Id
+    )
+    and cast(Value as varbinary(max)) <> @blankimage
```
Comment (pmamut) [inline one-off/nuke-workitem-embedded-images.sql:117]: so also update affected `EmbeddedImage.ContentType` and `EmbeddedImage_Now.ContentType` to `@imgpng`

File: one-off/nuke-workitem-embedded-images.sql [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	FETCH NEXT FROM cRichTextFields INTO @LongStringID, @Value
+
+	WHILE @@FETCH_STATUS = 0
+	BEGIN
+		update Blob 
+		set Value = @blankimage, ContentType=@imgpng
+		where Hash in (
+			SELECT Hash
+			FROM dbo.ExtractHashesFromUrls(@Value)
+		)
```
Comment (pmamut) [inline one-off/nuke-workitem-embedded-images.sql:144]: welp, `Blob.Hash` is now a lie

i think we need to
1. create a new `Blob` with `@blankimage` value and matching (pre-computed) hash
2. delete every `Blob` with doomed `Hash` from the extracted list
3. rewrite original `LongString.Value`, replacing doomed hashes with the known blank image hash

File: one-off/nuke-workitem-embedded-images.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmamut) [reply one-off/nuke-workitem-embedded-images.sql]: welp, `Blob.ContentType` didn't exist until S-41793, when we started storing naked blobs to hold images being embedded while their asset was still being created in the UI, and needed a place to hold their content type

normally, `ContentType` is an attribute of the actual "storage" asset (`Attachment`, `Image`, etc), not an attribute of the asset's `Content`

> This defeats being able to redact the blob in-situ and not having to touch any referencer

this use case never really came up originally. also, pushing content type from existing assets into their blob contents would require extra `Blob` join any time we need to query for it, even when we aren't fetching the content itself.

File: one-off/nuke-workitem-embedded-images.sql (hunk 6 above)
Comment (pmamut) [reply one-off/nuke-workitem-embedded-images.sql]: > don't think Blob.Hash needs to actually correlate to the content.

true, it does not _currently_, as the hash value is being treated as an opaque unique blob ID

> I have a branch in progress somewhere that replaces it with an random value

i know we discussed it before on another PR somewhere. i'd prefer for `Hash` to stay as a true hash, if (when?) we ever decide to de-dupe the `Blob` values like we now do for `LongString`

Comment (pmamut) [reply one-off/nuke-workitem-embedded-images.sql]: > We could dedupe using another true hash, maybe an indexed calculated column, calculated db-side

why, though? existing value already _is_ a true `sha1` hash of the data

> clients that use the current `Hash` like an ID have to follow an upload with a query to fetch it

not sure what you mean, the hash is computed on-the-fly during blob upload, and its hashed URL is returned in upload response

File: one-off/nuke-workitem-embedded-images.sql (hunk 5 above)
Comment (pmamut) [reply one-off/nuke-workitem-embedded-images.sql]: on second thought, we probably ought to whack these blobs as well, and update their referencing `EmbeddedImage.Content` and `EmbeddedImage_Now.Content` with `ID` of the "blank image" blob

**Review by pmamut — APPROVED** [review]
LGTM :shipit:

**Pull request merged**